### PR TITLE
Scoping of changing / deletion of relations

### PR DESCRIPTION
**Pull request by stefanmajoor**

### CODE REVIEW

File: binder/views.py [hunk 1]
```diff
@@ -1802,20 +1805,21 @@ def post(self, request, pk=None):
...
 			raise BinderMethodNotAllowed()
 
-		# FIXME: ugly workaround, remove when Django bug fixed
-		# Try/except because https://code.djangoproject.com/ticket/27005
-		try:
-			if request.body not in (b'', b'{}'):
-				raise BinderRequestError('{}DELETE body must be empty or empty json object.'.format('UN' if undelete else ''))
-		except ValueError:
-			pass
+		if not skip_body_check:
```
[the remark this answers is not in the document]
Comment (stefanmajoor) [reply binder/views.py]: Just for my information, how would this affect existing code? It is fully backwards compatible. Or am I missing something?

I do agree with the fact that it is not really a nice change. I did consider the change you proposed as well during implementation, but I could not defend the implicit logic you are adding to the function. Therefore I took the approach to be explicit rather than implicit about what the function is doing.

So, what doesn't make sense to me is that with your proposal you are allowed to DELETE with an non-empty body if you do a PUT/PATCH request, but not if you do a DELETE request. If you would call this function somewhere else in your application, (e.g. a custom endpoint) this wouldn't make sense at all, and can cause weird bugs.

The real problem here is that there is not the correct level of abstraction in the Binder views, and that I can't delete a model with all the logic attached to it, without "mocking" a delete request. I don't think we are going to solve this problem now though, so apart of your and mine solution, there are two other solutions we might consider:

- Remove the empty body check all together. If I understand correctly, this is just a sanity check which warns the frontenders that they are doing something wrong
- Move this code to the dispatch method, which then attaches this logic to the actual "DELETE" http method. This will have some major consequences for some of our projects which use DELETE endpoints, with logic depending on the body. (E.g. deleting a list of objects)

Maybe @Daanvdk Can give his two cents about it

[the remark this answers is not in the document]
Comment (sjamaan) [reply binder/views.py]: > Just for my information, how would this affect existing code? It is fully backwards compatible. Or am I missing something?

It wouldn't be if someone had copy/paste overridden the `delete` method (which wouldn't accept that new keyword).

But I fully agree the abstraction is all wrong here.

File: binder/views.py [hunk 2]
```diff
@@ -1230,31 +1232,32 @@ def _store_m2m_field(self, obj, field, value):
...
-						method()
+						method(request)
 					except BinderValidationError as bve:
 						validation_errors.append(bve)
 				elif obj_field.field.null:
+					# A bit of a hack. We make a check, so we need to make sure that we do have the permissions
+					# for this
+					from binder.permissions.views import PermissionView
+					if isinstance(rmobj_view, PermissionView):
+						rmobj_view.scope_change(request, rmobj, {obj_field.field.name: None})
```
[the remark this answers is not in the document]
Comment (stefanmajoor) [reply binder/views.py]: Either way this is not a situation we haven't had before, because the actual logic isn't delegated to the subview. Personally, I think the best solution would be to not do the `setattr(` in this view, but use the `_store` method of the `rmobj_view` for instance. I however couldn't come up with a solution doing this which wouldn't have major side-effects.

The solution I was considering was to add the `scope_change` to the `ModelView` as well, which would solve the problem. I am not opposed do doing it this way. However, I felt like this solution  would contain the effects of this edge case much better. What is your opinion about this?

File: tests/__init__.py [hunk 3]
```diff
@@ -105,3 +118,10 @@
 connection = connections['default']
 executor = MigrationExecutor(connection)
 cmd.sync_apps(connection, executor.loader.unmigrated_apps)
+
+# Hack to make the view_country permission, which doesn't work with the MigrationCommand somehow
+from django.contrib.auth.models import Group, Permission, ContentType
```
[the remark this answers is not in the document]
Comment (stefanmajoor) [reply tests/__init__.py]: I don't know why the issue is necessary either, but it seems like the hack for the migrations doesn't create the necessary permissions as well.

These models are not per se part of the TestApp though, and therefore are not necessary in the testapp/admin and testapp/urls. They are only imported here to bring the database in a correct state.

File: tests/test_permission_view.py [hunk 4]
```diff
@@ -372,3 +375,253 @@ def test_multiput_existing_and_new_resource_through_main_endpoint(self):
...
+
+        res = self.client.put('/country/', data=jsondumps({
+            'data': [{
+                'id': country.pk,
+                'name': 'Nederland',
+                'cities': [city1.pk]
+            }]
+        }))
+
+        # This is not ok
```
[the remark this answers is not in the document]
Comment (stefanmajoor) [reply tests/test_permission_view.py]: Removed it all together, since it is clear from the code that it is ok
